### ISSUE TITLE
Add set_grad() method to be able to update a global model from local models using an A3C implementation

### ISSUE THREAD

**Issue by suitable-name** [LaurentMazare/tch-rs#951]
Hi, 

thanks for the great effort that already went into this project!
I was trying to implement an A3C algorithm from scratch using tch-rs. I had a problem updating the global model from the local models. When calling zero_grads() on the optimizer (rmsprop) the gradients where undefined, so I couldn't call the step() method without making a dummy pass first to initialize the gradients of the global model again. My solution was to add a "set_grad" method:
1. added following in tch-sys/libtch/torch\_api.h:
`    const char *atg_set_grad(tensor self, tensor grad);`

2. added following in tch-sys/libtch/torch\_api.cpp:
```
    // Wrapper function to expose gradient setting via C API
    const char *atg_set_grad(tensor self, tensor grad)
    {
      PROTECT(
          // Cast opaque C pointers back to C++ at::Tensor pointers
          auto self_tensor_ptr = reinterpret_cast<at::Tensor *>(self);
          auto grad_tensor_ptr = reinterpret_cast<at::Tensor *>(grad);
    
          // Basic null pointer check for safety
          if (!self_tensor_ptr) {
            throw std::runtime_error("Received null self tensor pointer in atg_set_grad");
          } if (!grad_tensor_ptr) {
            // Assume grad must be a valid tensor.
            throw std::runtime_error("Received null grad tensor pointer in atg_set_grad");
          }
    
          // The assignment operator for at::Tensor handles the reference counting (shared_ptr).
          self_tensor_ptr->mutable_grad() = *grad_tensor_ptr;
    
          // Return nullptr on success, indicating no C++ exception occurred.
          return nullptr;)
    }
```

3. added following in torch-sys/src/lib.rs (in 'extern "C"' block):
`    pub fn atg_set_grad(self_: *mut C_tensor, grad: *mut C_tensor) -> *const libc::c_char;`

4. added following in src/wrappers/tensor.rs (in 'impl Tensor' block):
```
        /// Sets the gradient of this tensor.
        ///
        /// This function assigns the provided `grad` tensor as the gradient for the `self` tensor.
        /// It performs checks to ensure compatibility (device, dtype, shape) and that `self`
        /// requires gradients.
        ///
        /// # Arguments
        ///
        /// * `grad` - The tensor to set as the gradient.
        ///
        /// # Errors
        ///
        /// Returns `TchError` if:
        /// * `self` does not require gradients.
        /// * `self` and `grad` are on different devices.
        /// * `self` and `grad` have different data types (Kind).
        /// * `self` and `grad` have different shapes.
        /// * An error occurs during the underlying C++ call.
        ///
        /// # Safety Note
        ///
        /// On success, the ownership of the `grad` tensor is effectively shared with the `self`
        /// tensor's internal gradient storage within LibTorch. The Rust `grad` Tensor object
        /// is forgotten using `std::mem::forget` to prevent a double-free or premature release.
        pub fn set_grad(&self, grad: Tensor) -> Result<(), TchError> {
            if !self.requires_grad() {
                return Err(TchError::Kind(format!(
                    "Cannot set gradient on tensor that does not require grad",
                )));
            }
    
            if self.device() != grad.device() {
                return Err(TchError::Kind(format!(
                    "Tensor device ({:?}) and gradient device ({:?}) mismatch",
                    self.device(),
                    grad.device()
                )));
            }
    
            if self.kind() != grad.kind() {
                return Err(TchError::Kind(format!(
                    "Tensor dtype ({:?}) and gradient dtype ({:?}) mismatch",
                    self.kind(),
                    grad.kind()
                )));
            }
    
            if self.size() != grad.size() {
                return Err(TchError::Kind(format!(
                    "Tensor shape ({:?}) and gradient shape ({:?}) mismatch",
                    self.size(),
                    grad.size()
                )));
            }
    
            let self_ptr = self.c_tensor;
            let grad_ptr = grad.c_tensor;
    
            unsafe {
                // Call the FFI function defined in tch-sys
                let error_ptr = torch_sys::atg_set_grad(self_ptr, grad_ptr);
    
                // Check the return value for errors
                if error_ptr.is_null() {
                    // Success: C++ side now shares ownership of the gradient tensor's data
                    // via its shared_ptr. Prevent Rust from dropping its Tensor handle
                    // for 'grad', which would decrement the shared_ptr count prematurely.
                    std::mem::forget(grad);
                    Ok(())
                } else {
                    Err(TchError::Torch(
                        "Failure: Convert the C error string to a TchError.".to_string(),
                    ))
                }
            }
        }
```

According to my logs the solution seems to work:
`    INFO ThreadId(05) rl::agents::a3c::agent: Successfully set gradient. dest_grad_defined=true`

I'm planning to also add an PR later, but for now, here is a solution, if anyone else has that problem.
